README.md
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.2616951.svg)](https://doi.org/10.5281/zenodo.2616951)
[![R-CMD-check](https://github.com/imbi-heidelberg/adoptr/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/imbi-heidelberg/adoptr/actions/workflows/R-CMD-check.yaml)
[![Codecov test coverage](https://codecov.io/gh/imbi-heidelberg/adoptr/branch/master/graph/badge.svg)](https://codecov.io/gh/imbi-heidelberg/adoptr?branch=master)
[![CRAN status](https://www.r-pkg.org/badges/version/adoptr)](https://cran.r-project.org/package=adoptr)
[![](http://cranlogs.r-pkg.org/badges/last-month/adoptr?color=green)](https://cran.r-project.org/package=adoptr)
[![](http://cranlogs.r-pkg.org/badges/grand-total/adoptr?color=green)](https://cran.r-project.org/package=adoptr)


# adoptr <a href='https://github.com/imbi-heidelberg/adoptr'><img src='man/figures/logo.png' align="right" height="139" /></a>


Adaptive optimal two-stage designs for clinical trials with one or two arms. 
For details on the core theoretical background, see:

> Pilz M, Kunzmann K, Herrmann C, Rauch G, Kieser M. A variational approach to
optimal two-stage designs. *Statistics in Medicine*. 2019;38(21):4159–4171.
https://doi.org/10.1002/sim.8291



## Installation

Install the latest version directly from [GitHub](https://github.com/) with:

```r
devtools::install_github("imbi-heidelberg/adoptr")
```



## Documentation

The documentation is hosted at https://imbi-heidelberg.github.io/adoptr.



## Validation Report

We provide an extensive validation report for **adoptr** which is implemented 
using the **bookdown** package.
The sources are available at https://github.com/imbi-heidelberg/adoptr-validation-report and
the last build version is hosted at https://imbi-heidelberg.github.io/adoptr-validation-report.


NEWS.md
# adoptr 1.1.0

* added log-rank test for time-to-event-endpoints
* added chi-squared distribution for contingency tables and two-sided tests
* added f-distribution for ANOVA
* improved get_initial_design to find better initial designs
* added automatic constraint checks
* bug-fix for binomial pdf/cdf
* bug-fix in design2str function

# adoptr 1.0.0

* added references to article in journal of statistical software

# adoptr 0.4.3

* corrected class diagram in vignette

# adoptr 0.4.2

* minor bug/print fixes
* Student-t data distribution
* switched to GitHub Actions for CI/CD

# adoptr 0.4.0

* included JSS article as vignette

# adoptr 0.3.2

* bugfix for binomial pdf/cdf

# adoptr 0.3.1

* added support for binomial endpoint
* initial design creation more convenient
* extended summary() function for two stage designs

# adoptr 0.3.0

* bugfix for pretty printing of optimization results
* streamlined code for internal evaluation of integral scores during optimization; 
substantial speedup!

# adoptr 0.2.3

* reworked the class printing system to be more informative
* minimize() now also returns an S3 class object to allow pretty-printing



# adoptr 0.2.2

* fixed references to SIM paper



# adoptr 0.2.1

* added references to SIM paper in all relevant places



# adoptr 0.2.0

* new feature: composite scores allows more generic expressions than just 
    affine score combinations, cf. composite()
* affine scores (s1 + 2*s2) are no longer supported, use composite instead
* more consistent class system: conditional scores no longer need a specification
    of distributions by default (no need for conditional sampel size e.g.).
    Instead, expected() now requires explicit specification of the data and
    prior distribution to integrate with.
* Vignettes updated
* fixed broked tests due to updated rpact



# adoptr 0.1.1

* extended Description field in DESCRIPTION to full paragraph
* provided examples for all user facing functions
* revision of docs



# adoptr 0.1.0

* initial release


tests/testthat/test_ANOVA.R
context("F-distribution")

test_that("Constructors work", {

    expect_true(
        NestedModels(4, 5)@p_inner == 4 &
        NestedModels(4, 5)@p_outer ==5
    )

    expect_true(
        ANOVA(6)@p_inner == 1 &
        ANOVA(6)@p_outer == 6
    )

    expect_error(
        NestedModels(2.3, 1)
    )

    expect_error(
        NestedModels(2, -1)
    )

    expect_error(
        ANOVA(3.4)
    )

    expect_error(
        ANOVA(-6)
    )
})

test_that("thetas and ncps are correctly computed", {


    # ANOVA
    p_vec <- c(0.8, 0.15, 0.3, 0.96)
    n <- 52
    sigma <- 2
    dist <- ANOVA(length(p_vec))

    real_ncp <- n * (sum((p_vec - mean(p_vec))^2)) / sigma^2
    ncp_calc <- (n * (dist@p_outer - dist@p_inner + 1)) * get_tau_ANOVA(p_vec, sigma)

    expect_equal(
        real_ncp, ncp_calc
    )

})

test_that("pdf is defined correctly", {

    dist <- ANOVA(3)
    x <- seq(0.1, 5, by = 0.1)
    n <- 42
    p_vec <- c(0.3, 0.4, 0.5)
    theta <- get_tau_ANOVA(p_vec)
    real_ncp <- n * (sum((p_vec - mean(p_vec))^2))
    expect_equal(
        probability_density_function(dist, x, n, theta),
        stats::df(x, 2, 3 * n - 3, ncp = real_ncp)
    )

})

test_that("cdf is defined correctly", {

    dist <- ANOVA(3)
    x <- seq(0.1, 5, by = 0.1)
    n <- 32
    p_vec <- c(0.2, 0.4, 0.5)
    theta <- get_tau_ANOVA(p_vec)
    real_ncp <- n * (sum((p_vec - mean(p_vec))^2))
    expect_equal(
        cumulative_distribution_function(dist, x, n, theta),
        stats::pf(x, 2, 3 * n - 3, ncp = real_ncp)
    )
})

test_that("quantile is defined correctly", {

    dist <- ANOVA(4)
    x <- seq(0.1, 5, by = 0.1)
    n <- 22
    p_vec <- c(0.2, 0.4, 0.7, 0.36)
    theta <- get_tau_ANOVA(p_vec)
    real_ncp <- n * (sum((p_vec - mean(p_vec))^2))
    expect_equal(
        cumulative_distribution_function(dist, x, n, theta),
        stats::pf(x, 3, 4 * n - 4, ncp = real_ncp)
    )
})

test_that("simulate respects seed", {

    expect_equal(
        simulate(ANOVA(3), 10, 10, 0.2, seed = 42),
        simulate(ANOVA(3), 10, 10, 0.2, seed = 42),
        tolerance = 1e-6, scale = 1)

    set.seed(42)

    expect_true(
        all(simulate(ANOVA(4), 10, 12, 1.1) != simulate(ANOVA(4), 10, 12, 1.1)))
})

test_that("show_method", {

    expect_equal(
        capture.output(show(NestedModels(4, 5))),
        "NestedModels<p_inner=4,p_outer=5> "
    )

    expect_equal(
        capture.output(show(ANOVA(4))),
        "ANOVA<n_groups=4> "
    )
})


tests/testthat/test_minimize.R
context("check minimize()")



# preliminaries
order <- 5L

null        <- PointMassPrior(.0, 1)
alternative <- PointMassPrior(.4, 1)
datadist    <- Normal(two_armed = FALSE)

ess   <- ExpectedSampleSize(datadist, alternative)
ess_0 <- ExpectedSampleSize(datadist, null)
cp    <- ConditionalPower(datadist, alternative)
pow   <- expected(cp, datadist, alternative)
toer  <- Power(datadist, null)

alpha <- 0.05
beta  <- 0.2

initial_design <- get_initial_design(.4, alpha, beta, "two-stage", dist = datadist, order = order)



test_that("nloptr maxiter warning correctly passed", {

    expect_warning(

        minimize(
            ess,
            subject_to(
                pow  >= 1 - beta,
                toer <= alpha
            ),
            initial_design,
            opts = list(
                algorithm   = "NLOPT_LN_COBYLA",
                xtol_rel    = 1e-5,
                maxeval     = 10
            )
        )
    )

})



test_that("nloptr invalid initial values error works", {

    lb_design    <- update(initial_design, c(5, -1, 2, numeric(order), numeric(order) - 3))
    lb_design@n1 <- initial_design@n1 + 1

    expect_error(

        minimize(
            ess,
            subject_to(
                pow  >= 1 - beta,
                toer <= alpha
            ),
            initial_design,
            lower_boundary_design = lb_design,
            opts = list(
                algorithm   = "NLOPT_LN_COBYLA",
                xtol_rel    = 1e-5,
                maxeval     = 10
            )
        )

    )

})


test_that("Optimal one-stage design can be computed", {

    opt_os <<- minimize(
        ess,
        subject_to(
            pow  >= 1 - beta,
            toer <= alpha
        ),
        initial_design = get_initial_design(.4, alpha, beta, "one-stage", dist=datadist, order=order)
    )

    expect_equal(
        opt_os$design@c1f,
        qnorm(1 - alpha),
        tolerance = sqrt(.Machine$double.eps), scale = 1)

    expect_equal(
        opt_os$design@n1,
        ((qnorm(1 - beta) + qnorm(1 - alpha)) / 0.4)^2,
        tolerance = 1e-4, scale = 1)

}) # end 'optimal one-stage design can be computed'



test_that("Optimal group-sequential design is computable", {

    initial_design_gs <- get_initial_design(.4, alpha, beta, "group-sequential", dist=datadist, order=order)

    opt_gs <<- minimize(
        ess,
        subject_to(
            pow  >= 1 - beta,
            toer <= alpha
        ),
        initial_design_gs
    )

    expect_equal(
        round(evaluate(pow, opt_gs$design), 1),
        0.8,
        tolerance = sqrt(.Machine$double.eps), scale = 1)

    expect_equal(
        round(evaluate(toer, opt_gs$design), 2),
        0.05,
        tolerance = sqrt(.Machine$double.eps), scale = 1)


    # Check if n2 is equal at boundaries
    expect_equal(
        n2(opt_gs$design, opt_gs$design@c1f),
        n2(opt_gs$design, opt_gs$design@c1e))

    expect_equal(
        opt_gs$nloptr_return$solution[1],
        opt_gs$design@n1)

}) # end 'optimal group-sequential design is computable'



test_that("Optimal group-sequential design is superior to standard gs design", {

    # Create design from rpact
    design_rp <- rpact::getDesignInverseNormal(
        kMax = 2,
        alpha = alpha,
        beta = beta,
        futilityBounds = 0,
        typeOfDesign = "P")
    res <- rpact::getSampleSizeMeans(
        design_rp, normalApproximation = TRUE, alternative = .4 * sqrt(2))
    c2_fun <- function(z){
        w1 <- 1 / sqrt(2)
        w2 <- sqrt(1 - w1^2)
        out <- (design_rp$criticalValues[2] - w1 * z) / w2
        return(out)
    }
    c1f <- qnorm(
        rpact::getDesignCharacteristics(design_rp)$futilityProbabilities
    ) + sqrt(res$numberOfSubjects1[1]) * .4
    rpact_design <- GroupSequentialDesign(
            ceiling(res$numberOfSubjects1[1,]),
        c1f,
        design_rp$criticalValues[1],
        ceiling(res$numberOfSubjects1[2,]),
        rep(2.0, 100),
        100L
    )
    rpact_design@c2_pivots <- sapply(adoptr:::scaled_integration_pivots(rpact_design), c2_fun)

    # use opt_gs from above
    testthat::expect_lte(
        evaluate(ess, opt_gs$design),
        evaluate(ess, rpact_design))

})



test_that("base-case satisfies constraints", {

    opt_ts <<- minimize(
        ess,
        subject_to(
            pow  >= 1 - beta,
            toer <= alpha
        ),
        initial_design
    )

    # compute summaries
    out <- summary(opt_ts$design, "power" = pow, "toer" = toer, "CP" = cp, rounded = FALSE)

    out2 <- summary(opt_ts$design, "power" = pow, "toer" = toer, rounded = FALSE)

    out3 <- summary(opt_ts$design,  "CP" = cp, rounded = FALSE)

    out4 <- summary(opt_ts$design, rounded = TRUE)

    expect_equal(
        as.numeric(out$uncond_scores["power"]),
        0.8,
        tolerance = 1e-3, scale = 1)

    expect_equal(
        as.numeric(out$uncond_scores["power"]),
        as.numeric(out2$uncond_scores["power"]),
        tolerance = 1e-3, scale = 1)

    expect_equal(
        as.numeric(out$uncond_scores["toer"]),
        0.05,
        tolerance = 1e-3, scale = 1)


    expect_equal(
        as.numeric(out$uncond_scores["toer"]),
        as.numeric(out2$uncond_scores["toer"]),
        tolerance = 1e-3, scale = 1)


}) # end base-case respects constraints



test_that("base-case results are consistent - no post processing", {

    # optimal two-stage design better than optimal group-sequential design
    expect_lt(
        evaluate(ess, opt_ts$design),
        evaluate(ess, opt_gs$design))

    # optimal group-sequential design better than optimal one-stage design
    expect_lt(
        evaluate(ess, opt_gs$design),
        evaluate(ess, opt_os$design))


    # simulate on boundary of null
    sim_null <- simulate(
        opt_ts$design, nsim = 10^6, dist = datadist, theta = .0, seed = 54)

    # check type one error rate on boundary of null
    expect_equal(
        mean(sim_null$reject),
        alpha,
        tolerance = 1e-3, scale = 1)

    # expected sample size on boundary of null
    expect_equal(
        mean(sim_null$n2 + sim_null$n1),
        evaluate(ess_0, opt_ts$design),
        tolerance = 1e-2, scale = 1)

    # simulate under alternative
    sim_alt  <- simulate(
        opt_ts$design, nsim = 10^6, dist = datadist, theta = .4, seed = 54)

    # check power constraint
    expect_equal(
        mean(sim_alt$reject),
        1 - beta,
        tolerance = 1e-2, scale = 1)

    # check expected sample size under alternative
    expect_equal(
        mean(sim_alt$n2 + sim_alt$n1),
        evaluate(ess, opt_ts$design),
        tolerance = .5, scale = 1)

    # maximum sample size of adaptive design is larger than of one-stage design
    mss <- MaximumSampleSize()
    expect_lte(
        evaluate(mss, opt_os$design),
        evaluate(mss, opt_ts$design)
    )

}) # end 'base-case results are consistent'




test_that("conditional constraints work", {

    opt_ts <- suppressWarnings(

        minimize( # ignore: initial design is infeasible
            ess,
            subject_to(
                pow  >= 1 - beta,
                toer <= alpha,
                cp   >= 0.75,
                cp   <= 0.95
            ),
            initial_design,
            opts = list(
                algorithm   = "NLOPT_LN_COBYLA",
                xtol_rel    = 1e-4,
                maxeval     = 10000
            )
        )
    )

    tol <- .005
    # check lower boundary on conditional power
    expect_gte(
        evaluate(cp, opt_ts$design, opt_ts$design@c1f),
        .75 - tol)

    # check lower boundary on conditional power
    expect_lte(
        evaluate(cp, opt_ts$design, opt_ts$design@c1e),
        .95 + tol)

    # test that c2 is monotonously increasing
    expect_true(all(
        sign(diff(opt_ts$design@c2_pivots)) == -1))

}) # end 'conditional constraints work'



test_that("conditional constraints work", {

    expect_equal(
        capture.output(print(opt_os)),
        "OneStageDesign<optimized;n=39;c=1.64> "
    )


}) # end 'conditional constraints work'




test_that("initial design works", {
    expect_error(
        get_initial_design(.4, .025, .2, "adaptive", dist = Normal(), order = 6L)
    )

    expect_error(
        get_initial_design(0.4, 0.025, 0.2, dist = Survival(0.8))
    )

    expect_error(
        get_initial_design(.4, 1.025, .2, "two-stage", dist = Normal(), order = 6L)
    )

    expect_true(
        is(get_initial_design(.4, .025, .2, "two-stage", dist = Normal(F), order = 6L), "TwoStageDesign")
    )

    expect_true(
        is(get_initial_design(1.4, .025, .2, "two-stage", type_n2 = "linear_decreasing", dist = Survival(0.7), order = 6L, slope = -10),
           "TwoStageDesignSurvival")
    )

    expect_true(
        is(get_initial_design(1.4, .025, .2, "two-stage", dist = Survival(0.7), order = 6L),
           "TwoStageDesignSurvival")
    )

    expect_true(
        is(get_initial_design(1.4, .025, .2, "two-stage", type_n2 = "linear_increasing", dist = Survival(0.7), order = 6L, slope = 10),
           "TwoStageDesignSurvival")
    )

    expect_true(
        is(get_initial_design(.4, .025, .2, "group-sequential", dist = Normal(), order = 6L), "GroupSequentialDesign")
    )

    expect_true(
        is(get_initial_design(.4, .025, .2, "group-sequential", dist=Normal(), type_c2 = "constant",
                              order = 6L), "GroupSequentialDesign")
    )

    expect_true(
        is(get_initial_design(1.4, .025, .2, "group-sequential", dist = Survival(0.7), order = 6L),
           "GroupSequentialDesignSurvival")
    )

    expect_true(
        is(get_initial_design(.4, .025, .2, "one-stage", dist = Normal(), order = 6L),
           "OneStageDesign")
    )

    expect_true(
        is(get_initial_design(0.04, 0.05, 0.2, "group-sequential", dist = ANOVA(2)),
           "GroupSequentialDesign")
    )

    expect_error(
        get_initial_design(0.4, 0.2, 0.8, ce = 1, cf = 2)
    )


    init <- get_initial_design(.4, .025, .2, "two-stage", dist = Normal(F), cf = 0.5,
                          type_c2 = "constant", ce = 2.5)

    #check values futility boundary
    expect_true(
        init@c1f == 0.5
    )

    # check that c2 is constant
    expect_true(
        all(init@c2_pivots-init@c2_pivots[1] == rep(0, 7))
    )

    #check efficacy boundary
    expect_true(
        init@c1e == 2.5
    )

    init2 <- get_initial_design(.4, .025, .2, "two-stage", dist = Normal(T),
                                type_n2 = "linear_increasing", slope = 20)

    #check that n2 is increasing
    expect_false(
        is.unsorted(init2@n2_pivots)
    )

    init2new <- get_initial_design(.4, .025, .2, "two-stage", dist = Normal(T),
                                type_n2 = "linear_increasing", info_ratio = 0.3)

    #check that n2 is increasing
    expect_false(
        is.unsorted(init2new@n2_pivots)
    )


    #check the slope
    expect_true(
        n2(init2,1.5)-n2(init2,0.5) == 20
    )

    init3 <- get_initial_design(.4, .025, .2, "two-stage", dist = Normal(T),
                                type_c2 = "constant", info_ratio = 0.2, ce = 2)

    #check information ratio
    expect_equal(4 * rep(init3@n1, 7), init3@n2_pivots)


    #check that student distribution uses normal distribution
    init4 <- get_initial_design(0.4, 0.025, 0.1, "one-stage", dist = Student())
    expect_equal(
        init4@c1e,
        1.96,
        tolerance = 1e-3
    )

    #check warnings are raised when unnecessarily specifying n2 or c2 type
    expect_warning(
        get_initial_design(0.4, 0.025, 0.2, "one-stage", type_c2 = "constant")
    )

    expect_warning(
        get_initial_design(0.4, 0.025, 0.2, "one-stage", type_n2 = "constant")
    )

    #check warning when ce is chosen too small

    expect_warning(
        get_initial_design(0.4, 0.025, 0.2, "two-stage", type_c2 = "constant", ce = 1)
    )

    expect_warning(
        get_initial_design(0.4, 0.025, 0.2, "two-stage", type_c2 = "linear_decreasing", ce = 1)
    )

    #check that ce can be chosen

    expect_equal(
        get_initial_design(0.3, 0.025, 0.2, type_c2 = "linear_decreasing", ce = 3)@c1e,
        3
    )

    init5 <- get_initial_design(.4, .025, .2, "two-stage", dist = Normal(F), cf = 0.5, type_n2 = "constant",
                       type_c2 = "constant", ce = 2.5)

    #check that c2 is constant
    expect_true(
            all(init@c2_pivots-init@c2_pivots[1] == rep(0,7))
        )

    #check that c2 is linear decreasing
    init6 <- get_initial_design(1.4, .025, .2, "two-stage", dist=Survival(0.7), cf = 0.5, type_n2 = "constant",
                                type_c2 = "linear_decreasing")

    expect_false(
        is.unsorted(rev(init6@c2_pivots))
    )

    #check linear decreasing n2 design
    init7 <- get_initial_design(0.4, .025, .2, "two-stage", cf = 0.5,
                                type_n2 = "linear_decreasing")

    expect_false(
        is.unsorted(rev(init7@n2_pivots))
    )

    expect_error(
        get_initial_design(0.4, .025, .2, "two-stage", cf = 0.5,
                           type_n2 = "linear_decreasing", slope = 10)
    )

    expect_error(
        get_initial_design(0.4, .025, .2, "two-stage", cf = 0.5,
                           type_n2 = "linear_increasing", slope = -10)
    )


    #check that slope is automatically adjusted
    expect_warning(
        get_initial_design(0.4, .025, .2, "two-stage", cf = 0.5,
                                 type_n2 = "linear_decreasing",slope = -100))

    expect_warning(
        get_initial_design(0.4, .025, .2, "two-stage", cf = 0.5,
                           type_n2 = "linear_increasing",slope = 100))



}) # end 'initial design works'


test_that("constraint checks are working", {
    init <- TwoStageDesign(25, 0.2, 2.5,
                           c(80, 77, 70, 61, 50, 36, 25),
                           c(2.2, 2.1, 1.8, 1.4, 0.9, 0.3, -0.1), order = 7L)

    expect_warning(
        opt_design <- minimize(ess, subject_to(toer <= 0.01, pow >= 0.95, cp >= 0.95,
                                               MaximumSampleSize() <= 70),
                               initial_design = init, check_constraints = TRUE,
                               opts = list(algorithm = "NLOPT_LN_COBYLA", xtol_rel = 1e-05, maxeval = 1))

    )

    expect_warning(
        opt_design <- minimize(ess,subject_to(toer <= 0.025, pow >= 0.8, MaximumSampleSize() <= pow,
                                              cp >= ConditionalSampleSize()),
                               initial_design = init, check_constraints = TRUE,
                               opts = list(algorithm = "NLOPT_LN_COBYLA", xtol_rel = 1e-05, maxeval = 1))
    )
}) # end 'constraint checks are working'


tests/testthat/test_ChiSquared.R
context("Chi-Squared distribution")

test_that('constructors work', {

    expect_true(
        ChiSquared(df = 10)@multiplier == 1
    )

    expect_error(
        ChiSquared(df = -1)
    )

    expect_error(
        ChiSquared(df = 2.3)
    )

    expect_true(
        Pearson2xK(5)@df == 4
    )

    expect_true(
        Pearson2xK(5)@multiplier == 1 / (Pearson2xK(5)@df + 1)
    )

    expect_error(
        Pearson2xK(-3)
    )

    expect_error(
        Pearson2xK(1.8)
    )

    expect_equal(
        ZSquared()@multiplier, ZSquared()@df + 1
    )

    expect_equal(
        ZSquared(FALSE)@multiplier, ZSquared()@df
    )
})

test_that('thetas and ncps are correctly computed', {

    # theoretical ncp and calculated ncp are equal
    # Pearson2xK
    p_vector <- c(0.8, 0.15, 0.3, 0.96)
    n <- 52
    dist <- Pearson2xK(length(p_vector))
    real_ncp <- n / length(p_vector) * (sum((p_vector - mean(p_vector))^2)) /
        (mean(p_vector) * (1 - mean(p_vector)))
    ncp_calc <- n / dist@multiplier * get_tau_Pearson2xK(p_vector) * 1 / length(p_vector)

    expect_equal(ncp_calc, real_ncp)

    #ZSquared
    dist2 <- ZSquared(TRUE)
    mu <- 0.4
    sigma <- 1.2
    expect_equal(
        get_tau_ZSquared(mu, sigma) * n / dist2@multiplier,
        (sqrt(n / 2) * mu/sigma)^2
    )

    dist3 <- ZSquared(FALSE)
    expect_equal(
        get_tau_ZSquared(mu, sigma) * n / dist3@multiplier,
        (sqrt(n) * mu/sigma)^2
    )
})

test_that('pdf is defined correctly', {

    dist <- Pearson2xK(3)
    x       <- seq(-3, 3, by = .1)
    n       <- 22
    p_vec   <- c(0.5, 0.6, 0.7)
    theta   <- get_tau_Pearson2xK(p_vec)
    expect_equal(
        probability_density_function(dist, x, n, theta),
        stats::dchisq(x, df = 2, ncp = n * (sum((p_vec - mean(p_vec))^2)) /
                          (mean(p_vec) * (1 - mean(p_vec)))),
        tolerance = 1e-6, scale = 1)
})

test_that('cdf is defined correctly', {
    dist <- Pearson2xK(3)
    x       <- seq(-3, 3, by = .1)
    n       <- 22
    p_vec   <- c(0.5, 0.6, 0.7)
    theta   <- get_tau_Pearson2xK(p_vec)
    expect_equal(
        cumulative_distribution_function(dist, x, n, theta),
        stats::pchisq(x, df = 2, ncp = n * (sum((p_vec - mean(p_vec))^2)) /
                          (mean(p_vec) * (1 - mean(p_vec)))),
        tolerance = 1e-6, scale = 1)
})

test_that('quantile is defined correctly', {

    dist <- Pearson2xK(3)
    x       <- seq(0.1, 1, by = .01)
    n       <- 22
    p_vec   <- c(0.5, 0.6, 0.7)
    theta   <- get_tau_Pearson2xK(p_vec)
    expect_equal(
        quantile(dist, x, n, theta),
        stats::qchisq(x, df = 2, ncp = n * (sum((p_vec - mean(p_vec))^2)) /
                          (mean(p_vec) * (1 - mean(p_vec)))),
        tolerance = 1e-6, scale = 1)
})

test_that('simulate respects seed', {

    expect_equal(
        simulate(Pearson2xK(3), 10, 10, 0.2, seed = 42),
        simulate(Pearson2xK(3), 10, 10, 0.2, seed = 42),
        tolerance = 1e-6, scale = 1)

    set.seed(42)

    expect_true(
        all(simulate(Pearson2xK(4), 10, 12, 1.1) != simulate(Pearson2xK(4), 10, 12, 1.1)))
})

test_that("show method", {

    expect_equal(
        capture.output(show(Pearson2xK(4))),
        "Pearson2xK<df=3> "
    )

    expect_equal(
        capture.output(show(ZSquared())),
        "ZSquared<df=1> "
    )

    expect_equal(
        capture.output(show(ChiSquared(5))),
        "ChiSquared<df=5> "
    )
})


R/adoptr.R
#' Adaptive Optimal Two-Stage Designs
#'
#' The \pkg{adoptr} package provides functionality to explore custom optimal
#' two-stage designs for one- or two-arm superiority tests. More than two arms
#' can be compared via chi-squared tests or ANOVA.
#' For more details on the theoretical background see
#' <doi:10.1002/sim.8291> and <doi:10.18637/jss.v098.i09>.
#' \pkg{adoptr} makes heavy use of the S4 class system.
#' A good place to start learning about it can be found
#' \href{http://adv-r.had.co.nz/OO-essentials.html}{here}.
#'
#'
#'
#' @section Quickstart:
#'
#' For a sample workflow and a quick demo of the capabilities, see
#' \href{https://imbi-heidelberg.github.io/adoptr/articles/adoptr.html}{here}.
#'
#' A more detailed description of the background and the usage of \pkg{adoptr}
#' can be found \href{https://imbi-heidelberg.github.io/adoptr/articles/adoptr_jss.html}{here}
#' or here <doi:10.18637/jss.v098.i09> .
#'
#' A variety of examples is presented in the validation report hosted
#' \href{https://imbi-heidelberg.github.io/adoptr-validation-report/}{here}.
#'
#'
#'
#' @section Designs:
#'
#' \pkg{adoptr} currently supports \code{\link{TwoStageDesign}},
#' \code{\link{GroupSequentialDesign-class}}, and \code{\link{OneStageDesign-class}}.
#'
#'
#'
#' @section Data distributions:
#'
#' The implemented data distributions are \code{\link{Normal}}, \code{\link{Binomial}},
#' \code{\link{Student}}, \code{\link{Survival}}, \code{\link{ChiSquared}} (including
#' \code{\link{Pearson2xK}} and \code{\link{ZSquared}}) and \code{\link{ANOVA}}.
#'
#'
#'
#'
#' @section Priors:
#'
#' Both \code{\link{ContinuousPrior}} and \code{\link{PointMassPrior}} are
#' supported for the single parameter of a \code{\link{DataDistribution}}.
#'
#'
#'
#' @section Scores:
#'
#' See \code{\link{Scores}} for information on the basic system of representing
#' scores.
#' Available scores are \code{\link{ConditionalPower}},
#' \code{\link{ConditionalSampleSize}}, \code{\link{Power}}, and
#' \code{\link{ExpectedSampleSize}}..
#'
#' @import methods
#' @docType package
#' @name adoptr
NULL


R/ChiSquaredDistribution.R
#' Chi-Squared data distribution
#'
#' Implements a chi-squared distribution. The classes \code{Pearson2xk}
#' and \code{ZSquared} are subclasses, used in two different situations.
#' \code{Pearson2xK} is used when testing k groups for homogeneity in
#' response rates. The null hypothesis is
#' \ifelse{html}{\out{r<sub>1</sub>}=...=\out{r<sub>k</sub>}}{\eqn{r_1=...=r_k}}, and the
#' alternative is that there exists a pair of groups with differing rates.
#' \code{ZSquared} implements the square of a normally distributed random variable
#' with mean \eqn{\mu} and standard deviation \eqn{\sigma^2}.
#'
#' @template DataDistributionTemplate
#'
#' @rdname ChiSquaredDataDistribution-class
#' @exportClass ChiSquared
setClass("ChiSquared", representation(
    df  = "numeric",
    multiplier = "numeric"),
    contains = "DataDistribution")


#' Pearson's chi-squared test for contingency tables
#'
#' @include DataDistribution.R
#'
#' @rdname Pearson2xK-class
#' @exportClass Pearson2xK
setClass("Pearson2xK", contains = "ChiSquared")


#' Distribution class of a squared normal distribution
#'
#' @include DataDistribution.R
#'
#' @rdname ZSquared-class
#' @exportClass ZSquared
setClass("ZSquared", contains = "ChiSquared")


#' @param df number of degrees of freedom
#'
#' @examples
#' datadist <- ChiSquared(df=4)
#'
#' @seealso see \code{\link{probability_density_function}} and
#'    \code{\link{cumulative_distribution_function}} to evaluate the pdf
#'    and the cdf, respectively.
#'
#' @rdname ChiSquaredDataDistribution-class
#' @export
ChiSquared <- function(df) {
    if (df < 0 || abs(df - round(df)) > sqrt(.Machine$double.eps))
        stop("The degrees of freedom must be natural numbers.")
    new("ChiSquared", df = df, multiplier = 1)
}


#' Pearson's chi-squared test for 2 x k contingency tables
#'
#' When we test for homogeneity of rates in a k-armed trial with binary endpoints, the test statistic is
#' chi-squared distributed with \eqn{k-1} degrees of freedom under the null. Under the alternative, the statistic is chi-squared distributed with a
#' non-centrality parameter \eqn{\lambda}. The function \code{get_tau_Pearson2xk} then computes \eqn{\tau}, such that \eqn{\lambda} is given
#' as \eqn{n \cdot \tau}, where \eqn{n} is the number of subjects per group. In \code{adoptr}, \eqn{\tau} is used in the same way as \eqn{\theta}
#' in the case of the normally distributed test statistic.
#'
#' @param n_groups number of groups considered for testing procedure
#'
#' @examples
#' pearson <- Pearson2xK(3)
#'
#'
#' @rdname Pearson2xK-class
#' @export
Pearson2xK <- function(n_groups) {
    if (n_groups < 0 || abs(n_groups - round(n_groups)) > sqrt(.Machine$double.eps))
        stop("The number of groups must be a natural number.")
    new("Pearson2xK", df = n_groups - 1L, multiplier = 1/n_groups)
}

#' @param p_vector vector denoting the event rates per group
#'
#' @examples
#' H1 <- PointMassPrior(get_tau_Pearson2xK(c(.3, .25, .4)), 1)
#'
#' @rdname Pearson2xK-class
#'
#' @export
get_tau_Pearson2xK <- function(p_vector) {
    n_groups <- length(p_vector)
    mean_p <- mean(p_vector)
    deltas <- p_vector - mean_p
    tau <- (sum(deltas^2)/n_groups) / (mean_p * (1-mean_p))
    tau
}


#' Squared normally distributed random variable
#'
#' Implementation of \eqn{Z^2}, where \eqn{Z} is normally distributed with mean \eqn{\mu} and variance
#' \eqn{\sigma^2}. \eqn{Z^2} is chi-squared distributed with \eqn{1} degree of freedom and non-centrality parameter \eqn{(\mu/\sigma)^2}.
#' The function \code{get_tau_ZSquared} computes the factor \eqn{\tau=(\mu/\sigma)^2}, such that
#' \eqn{\tau} is the equivalent of \eqn{\theta} in the normally distributed case.
#' The square of a normal distribution \eqn{Z^2} can be used for two-sided hypothesis testing.
#'
#' @param two_armed logical indicating if a two-armed trial is regarded
#'
#' @examples
#' zsquared <- ZSquared(FALSE)
#'
#'
#' @rdname ZSquared-class
#' @export
ZSquared <- function(two_armed = TRUE) {
    new("ZSquared", df = 1L, multiplier = 1L + two_armed)
}


#' @param mu mean of Z
#' @param sigma standard deviation of Z
#'
#' @examples
#' H1 <- PointMassPrior(get_tau_ZSquared(0.4, 1), 1)
#'
#' @rdname ZSquared-class
#'
#' @export
get_tau_ZSquared <- function(mu, sigma = 1){
    (mu/sigma)^2
}


#' @examples
#' probability_density_function(Pearson2xK(3), 1, 30, get_tau_Pearson2xK(c(0.3, 0.4, 0.7, 0.2)))
#' probability_density_function(ZSquared(4), 1, 35, get_tau_ZSquared(0.4))
#'
#'
#' @rdname probability_density_function
#' @export
setMethod("probability_density_function", signature("ChiSquared", "numeric", "numeric", "numeric"),
          function(dist, x, n, theta, ...) {
              return(stats::dchisq(x, df = dist@df, ncp = n / dist@multiplier * theta))
          })


#' @examples
#' cumulative_distribution_function(Pearson2xK(3), 1, 30, get_tau_Pearson2xK(c(0.3,0.4,0.7,0.2)))
#' cumulative_distribution_function(ZSquared(4), 1, 35, get_tau_ZSquared(0.4))
#'
#'
#' @rdname cumulative_distribution_function
#' @export
setMethod("cumulative_distribution_function", signature("ChiSquared", "numeric", "numeric", "numeric"),
          function(dist, x, n, theta, ...) {
              return(stats::pchisq(x, df = dist@df, ncp = n / dist@multiplier * theta))
        })


#' @param probs vector of probabilities
#' @rdname ChiSquaredDataDistribution-class
#' @export
setMethod("quantile", signature("ChiSquared"),
          function(x, probs, n, theta, ...) { # must be x to conform with generic
              return(stats::qchisq(probs, df = x@df, ncp = n / x@multiplier * theta))
          })



#' @rdname ChiSquaredDataDistribution-class
#'
#' @param object object of class \code{ChiSquared}
#' @param nsim number of simulation runs
#' @param seed random seed
#'
#' @export
setMethod("simulate", signature("ChiSquared", "numeric"),
          function(object, nsim, n, theta, seed = NULL, ...) {
              if (!is.null(seed)) set.seed(seed)
              return(stats::rchisq(nsim, df = object@df, ncp = n / object@multiplier * theta))
})

setMethod("print", signature('ChiSquared'), function(x, ...) {
    glue::glue(
        "{class(x)[1]}<df={x@df}>"
    )
})
